util/topgen.py

```diff
@@ -818,12 +818,13 @@ def generate_top_ral(top, ip_objs, out_path):
             mem.n_bits = top_block.width
             top_block.wins.append(mem)
 
-    # get sub-block base addresses from top cfg
-    for block in top_block.blocks:
+    # get sub-block base addresses, instance names from top cfg
+    for ip in top_block.ips:
         for module in top["module"]:
-            if block.name == module["name"]:
-                block.base_addr = int(module["base_addr"], 0)
-                break
+            if ip.name == module["type"]:
+                ip.instance = module["name"]
+                ip.base_addr = int(module["base_addr"], 0)
+                top_block.blocks.append(deepcopy(ip))
```

Comment:
I dont see a need to maintain two lists. `blocks` can be a list of ALL instances of ALL IPs.

Reply:
@sriyerg 
let's take below as an example. If we have 2 EDN in the `blocks`, it will import `edn_ral_pkg` twice unless we use a separate list like `ips` or create a flag to indicate the EDN pkg has been imported. same for creating edn_ral_class.
```
% for i in block.blocks:
import ${i.name}_ral_pkg::*;
% endfor
```